### ISSUE TITLE
Query for unicode range \u036E-\u036F returns non-matching results

### ISSUE THREAD

**Issue by ZeLonewolf** [drolbr/Overpass-API#688]
The following query for a range of two consecutive unicode values returns 5,747 city nodes, however, none of the returned results actually appear to contain either character.

```
[out:csv(::id, name)][timeout:2500];
node[place=city][name~"[\u036E-\u036F]"];
out;
```
Queries for each character individually each return zero results:

```
[out:csv(::id, name)][timeout:2500];
node[place=city][name~"\u036E"];
out;
```

```
[out:csv(::id, name)][timeout:2500];
node[place=city][name~"\u036F"];
out;
```

**Comment by mmd-osm** [comment]
#332 is probably related...

**Comment by 1ec5** [comment]
[A query for `node[place=city][name~"[\u1ebf]"]`](https://overpass-turbo.eu/s/1tNT) (with just one backslash) does return [two](https://www.openstreetmap.org/node/369487050) [cities](https://www.openstreetmap.org/node/3140507587) that contain this combining character (because editors and imports at the time didn’t normalize the text to NFC). [Expanding the range](https://overpass-turbo.eu/s/1tNU) to U+0300 to U+036F correctly returns [this node](https://www.openstreetmap.org/node/369500682).

**Comment by 1ec5** [comment]
Oh, I just got lucky because the city names happened to contain some of the letters in the hexadecimal numbers in the range. Never mind me.

**Comment by mmd-osm** [comment]
So based on U+1EBF, I'm getting the following three place=city nodes (with proper unicode regex support):

```
  <node id="369487050"/>
  <node id="369487099"/>
  <node id="3140507587"/>
```

**Comment by ZeLonewolf** [comment]
I note that even with the escaping fixed, I still get (different) non-sensical results:
```
[out:csv(::id, name)][timeout:2500];
node[place=city][name~"[\\u036E-\\u036F]"];
out;
```

**Comment by mmd-osm** [comment]
Right, I've noticed the missing backslash when revisiting #332. In the end it doesn't make a whole lot of a difference, since the underlying regular expression implementation doesn't handle ranges as expected. 

I hope you received some link to a github gist to try out another implementation that works a bit better.

**Comment by NeatNit** [comment]
I've encountered this issue myself. It's baffling that character ranges can match characters outside the range.. How does this happen?

**Comment by NeatNit** [comment]
I remembered this bug and thought I knew why it happened, but I was wrong. Still, I did try to dig a little deeper and I want to share this minimal example:

```
[out:csv(::count,::id,name;true)];
rel(6195356);
map_to_area->.il; // Israel & Palestine

rel(area.il)[admin_level=8][name~"[\u06FE-\u06FF]"]->.x;
.x out count; // 8
rel.x[name~"\u06FE"]; out count; // 0
rel.x[name~"\u06FF"]; out count; // 0
.x out;
```

Output ([overpass turbo](https://overpass-turbo.eu/s/1ZUe)):
```
@count	@id	name
8	0	
0	0	
0	0	
	1381564	أبو غوش‎
	1382494	תל־אביב–יפו
	1383756	الشبلي-أم الغنم‎
	1386851	طمرة‎
	1387994	شعب‎
	1398018	زيمر‎
	1405261	كفر قاسم‎
	1933846	كعبية طباش حجاجرة‎
```

The range `[\u06FE-\u06FF]` contains only two characters, and neither of these characters actually appears in any name. And yet, the query finds 8 relations.

Edit: I'll also mention that the note @mmd-osm gave about escaping the backslash is wrong (though it may have been true back then). Run this query in any English-speaking area and witness the difference:
```
[out:csv(::count,name;true)];
way({{bbox}})[name~"[\\u0061-\\u007A]"];
out count;
way({{bbox}})[name~"[\u0061-\u007A]"];
out count;
way({{bbox}})[name~"[a-z]"];
out count;
(way({{bbox}})[name~"[\\u0061-\\u007A]"]; - way({{bbox}})[name~"[a-z]"];);
out count;
out;
```
It seems `"[\\u0061-\\u007A]"` is interpreted as: any of the characters \\,u,0,6,7,A and any of the characters in the range 1 to \\. The range 1-\\ includes all the digits (apart from 0), some symbols, and all the capital letters in ASCII.

However there is a chance that this is Overpass Turbo doing something weird, and not the Overpass API. I did not test outside of Turbo.

**Comment by mmd-osm** [comment]
Right, escaping does not help here, results are still unpredictable. In the end you would need a regex engine which fully supports unicode ranges. I've already mentioned that like 2 years ago.

On the other hand, if you don't do any escaping, your string will be converted much earlier, before it reaches the regular expression engine. The following to examples are exactly the same from a regular expression engine point of view. 

```
way({{bbox}})[name~"[\u0061-\u007A]"];
way({{bbox}})[name~"[a-z]"];
```

You can try this out using the first godbolt link I've posted below, or the following QL query:

```
make x bla = "\u0061-\u007A",
       bla2 = "\\u0061-\\u007A";
out;
```

- https://cpp.godbolt.org/z/jKE3WWbzr (regex.h like Overpass API, does not work)
- https://cpp.godbolt.org/z/zoTGYhP3E (std::regex, does not work)
- https://cpp.godbolt.org/z/1s15KxsYc (pcre2, works, see screenshot below)


![Image](https://github.com/user-attachments/assets/8f5f9db0-1e7d-4c75-9d8f-e3bc6e012d8c)

**Comment by NeatNit** [comment]
> In the end you would need a regex engine which fully supports unicode ranges. I've already mentioned that like 2 years ago.

Right, this is definitely the issue. So how do I get that fixed? Do I need to make a PR to switch to a Unicode regex library? 

Regarding the rest of your reply, I will have to look more costly at your examples, but just responding to the first one - I would expect that `\uXXXX` *in a C++ string literal* is replaced with the corresponding character, but I did not assume this happens in Overpass, more specifically in regular expressions in Overpass. From my perspective there is every possiblity that Overpass deliberately ignores backslashes in regular expressions, i.e. strings preceded by `~`, the same way Python does not parse backslashes in strings preceded by `r` such as `r"raw\nstring"`.

Now, I don't *assume* that it does, but I don't know that it doesn't, and in fact this may be a good idea for a feature if it isn't already.

I'm guessing that the reason the regex `"[\\u0061-\\u007A]"` didn't work is that the regex engine used in Overpass actually doesn't support `\uXXXX`. If so, it's woefully inadequate.

**Comment by NeatNit** [comment]
> * (pcre2, works, see screenshot below)

I don't know exactly what you were trying to show here. The screenshot doesn't work - Overpass complains of a regex syntax error. You don't actually show it working in the screenshot.

The godbolt link works though. I modified it by adding the `PCRE2_ALT_BSUX` compile option, which allows `\uXXXX` unicode escapes: https://cpp.godbolt.org/z/c9feMofez and, in my humble opinion, this library is what Overpass should be using.

**Comment by mmd-osm** [comment]
Note the “mmd dev” instance on the header, this is my fork not the official instance! It runs on pcre2. Check the object count on the bottom right, that’s the query result.

**Comment by mmd-osm** [comment]
> I modified it by adding the PCRE2_ALT_BSUX compile option, which allows \uXXXX unicode escapes

Good point, I need to check if this is relevant for https://github.com/mmd-osm/Overpass-API/blob/test7591/src/overpass_api/data/regular_expression.h#L291

**Comment by NeatNit** [comment]
> Note the “mmd dev” instance on the header, this is my fork not the official instance! It runs on pcre2. Check the object count on the bottom right, that’s the query result.

Right, I missed that! Looks like you also disabled the message "This map intentionally left blank. <small>(received empty dataset)</small>" that I have come to rely on.

> > I modified it by adding the PCRE2_ALT_BSUX compile option, which allows \uXXXX unicode escapes
> 
> Good point, I need to check if this is relevant for https://github.com/mmd-osm/Overpass-API/blob/test7591/src/overpass_api/data/regular_expression.h#L291

Great! I have yet to look at all the available compile options, so there may be more desirable flags, like `PCRE2_EXTRA_ALT_BSUX` which "in addition it recognizes \u{hhh..} as a hexadecimal character code, where hhh.. is any number of hexadecimal digits".

Do you intend to make a pull request out of this? If so, any idea when that will happen? For the majority of users, we can only rely on pubic instances and can't run our own, so using a proper Unicode regex library should IMO be pushed upstream for all Overpass users to benefit.

**Comment by mmd-osm** [comment]
> Do you intend to make a pull request out of this? 

 I have no plans to create a PR for it at this time.

> If so, any idea when that will happen? 

No idea, really. Without buy in from the maintainer, I'm afraid this topic won't move anywhere. I also don't have the bandwidth to make my fork available to a larger audience.

**Comment by NeatNit** [comment]
I see. @drolbr your thoughts? Overpass needs to use a regex engine with proper Unicode support. @mmd-osm has already done that in his fork (multiple different engines in fact). Would you be able (and willing) to replicate it in the main repo?

The current regex engine has very serious bugs with Unicode.

**Comment by 1ec5** [comment]
As a workaround, [QLever](https://qlever.cs.uni-freiburg.de/osm-planet/l1LP2e) does support Unicode escape sequences. However, its copy of OSM tends to lag by a couple weeks (sometimes a little more, but just [ping them](https://github.com/ad-freiburg/qlever/discussions) and they’ll restart the process). It would be nice to have the Overpass API as another option for tasks that require Unicode character classes.

**Comment by drolbr** [comment]
Back to the issue from [community.osm.org](https://community.openstreetmap.org/t/how-to-get-an-overpass-bug-fixed-regex-unicode-ranges/127058).

I'm searching for a regex engine that is intentionally implemented with [DFA](https://swtch.com/~rsc/regexp/regexp1.html) only and will stay so. Otherwise is practically sure that sooner or later someone will overuse the public instances, because the resource consumption of the regex engine is hard to predict.

There are in addition the requirements to be Open Source and long term maintained, as this so far I have not seen yet.
The algorithm is in fact not that complicated (in comparison with various input hardening exercises), and implementing an engine would allow to fine-control resource usage. Thus even implementing an engine on my own is not completely off the table, but rather not the first choice - uncontrolled input and a complex algorithm are a dangerous combination.

**Comment by NeatNit** [comment]
> I'm searching for a regex engine that is intentionally implemented with [DFA](https://swtch.com/~rsc/regexp/regexp1.html) only and will stay so.

Fascinating read. I'm still reading through it, but figured I could already ask - does the current regex.h library tick those boxes?

**Comment by 1ec5** [comment]
> I'm searching for a regex engine that is intentionally implemented with [DFA](https://swtch.com/~rsc/regexp/regexp1.html) only and will stay so. Otherwise is practically sure that sooner or later someone will overuse the public instances, because the resource consumption of the regex engine is hard to predict.

If you’re primarily concerned about controlling resource consumption, maybe you’d be interested in supporting a PEG such as [Lua’s pattern matching syntax](https://www.lua.org/pil/20.2.html). It’s considerably less expressive than POSIX but, by the same token, much simpler and controllable. Several standalone [C](https://github.com/dressupgeekout/libpatterns/) and [C++](https://github.com/PG1003/lex/) implementations are available as open source libraries. (There’s one in httpd as well.) Given the difference in capabilities, it might be a good idea to reserve a distinct flag or operator for patterns.

That said, I don’t think anything derived from Lua is going to come with suitable Unicode support either.

**Comment by waldyrious** [comment]
Just for completeness, I'll mention the [RE2 engine](https://github.com/google/re2), which if I understand correctly is [based on finite automata](https://opensource.googleblog.com/2010/03/re2-principled-approach-to-regular.html) (← written by Russ Cox, who the author of the article linked above), and has [support for named Unicode character classes](https://github.com/google/re2/wiki/syntax), but unfortunately [not arbitrary Unicode ranges](https://github.com/google/re2/issues/91).

That said, perhaps [Rust's `regex` crate](https://github.com/rust-lang/regex) could be an option? It is also based on DFA, has [good Unicode support](https://docs.rs/regex/latest/regex/#unicode-features), and does have a [C API](https://github.com/rust-lang/regex/tree/master/regex-capi).

**Comment by mmd-osm** [comment]
PCRE2 has a DFA only engine, which is slower and doesn't support JIT. For the JIT based approach I implemented, there are stack size and recursion limits in place (by PCRE2). Besides, the OS can kill the  `interpreter` process in case of excessive CPU or memory usage, which is a good thing in general.

https://wiki.openstreetmap.org/wiki/User:Mmd/Overpass_API_-_experimental_corner has some examples w/ PCRE2 JIT enabled by default, in case anyone wants to try their own examples and report back. Use the steering wheel on the right, and log on to your osm account in overpass turbo. Don't just copy the queries over to overpass-turbo.eu, this won't work.

-> https://wiki.openstreetmap.org/wiki/User:Mmd/Overpass_API_-_experimental_corner#Emojis_in_tag_values_(global)

**Comment by drolbr** [comment]
Thank you, RE2 looks promising.
